File: xgbDirect-hyperopt.R
suppressMessages (library (caret))
suppressMessages (library (lubridate))
suppressMessages (library (dplyr))
suppressMessages (library (xgboost))

source ("support.R")

fitFunc <- function (fitFormula, trainData, params)
{
       
        set.seed (4322)
        fit <- xgb.cv(params, booster="gbtree", objective="reg:linear", eval_metric="rmse", 
                      data=trainData, nround = 10000, nfold = 5, early.stop.round = 5,
                      verbose=FALSE)
        
        return (fit$test.rmse.mean[length (fit$test.rmse.mean)])
}

train.df <- read.csv ("data/train.csv")

train.df <- formatData (train.df, logTransform = TRUE) %>% tbl_df()
train.df$month <- factor (train.df$month)
train.df$year <- factor (train.df$year)

outlierDateTimes <- ymd_hms ("2012-11-07 00:00:00", "2011-01-09 22:00:00")
train.df <- filter (train.df, !datetime %in% outlierDateTimes)

train.formula <-  ~ season + holiday + workingday + weather + temp + atemp +
        humidity + windspeed + year + month + wday + day + hour - 1

trainData <- xgb.DMatrix (model.matrix (train.formula, train.df), label=train.df$registered)

# params <- list (eta=0.18542623350463372,
#                 gamma=2.072654565080585,
#                 max_depth=4,
#                 min_child_weight=0.8584180048511384,
#                 subsample=0.8482345927989308,
#                 colsample_bytree=0.7455594667162986)

#fit <- fitFunc (train.formula, trainData, params)

File: xgbDirect-hyperopt.py
import rpy2.robjects as ro
import pandas as pd
from rpy2.robjects import pandas2ri
from hyperopt import fmin, tpe, hp

def objective (params):
    eta, max_depth, subsample, col_sample_bytree, min_child_weight, gamma = params

    print "eta: ", eta
    rParams = ro.ListVector ({
        'eta': ro.FloatVector ([eta]),
        'max_depth': ro.IntVector ([max_depth]),
        'subsample': ro.FloatVector ([subsample]),
        'col_sample_bytree': ro.FloatVector ([col_sample_bytree]),
        'min_child_weight': ro.FloatVector ([min_child_weight]),
        'gamma': ro.FloatVector ([gamma])})

    f1=open ('./hyperopt-results', "a")
    
    print "--------------New Trial---------------"
    print "eta: ", eta
    print "max_depth: ", max_depth
    print "subsample: ", subsample
    print "col_sample_bytree: ", col_sample_bytree
    print "min_child_weight: ", min_child_weight
    print "gamma: ", gamma

    print >> f1, "--------------New Trial---------------"
    print >> f1, "eta: ", eta
    print >> f1, "max_depth: ", max_depth
    print >> f1, "subsample: ", subsample
    print >> f1, "col_sample_bytree: ", col_sample_bytree
    print >> f1, "min_child_weight: ", min_child_weight
    print >> f1, "gamma: ", gamma

    ro.globalenv['params'] = rParams
    rmse = ro.r('fitFunc (fitFormula, trainData, params)')
    print (rmse[0])
    print >> f1, "Result is: ", rmse[0]
    f1.close()
    return (rmse[0])
    
    
space = (
    hp.loguniform ('eta', -6, 0),
    hp.randint ('max_depth', 6) + 2,
    hp.uniform ('subsample', 0.5, 1),
    hp.uniform ('col_sample_bytree', 0.5, 1),
    hp.uniform ('min_child_weight', 0, 3), 
    hp.uniform ('gamma', 0, 3)
    )
    
pandas2ri.activate()

ro.r('source ("xgbDirect-hyperopt.R")')
       
best = fmin (objective,
             space = space,
             algo = tpe.suggest,
             max_evals= 1)

print best
            

